### PR TITLE
Implement Store, BaseQuery and ResultSet

### PR DESCRIPTION
**Pull request by erizocosmico**
`model` and `id` and `schema` are not final implementations. I needed the actual implementation of them in order for the rest of stuff to work, but the final implementation will be the one from @dpordomingo

### CODE REVIEW

File: store.go [hunk 1]
```diff
@@ -1,199 +1,207 @@
...
 }
 
-// Insert insert the given document in the collection, returns error if no-new
-// document is given. The document id is setted if is empty.
-func (s *Store) Insert(doc DocumentBase) error {
-	if !doc.IsNew() {
+// Insert insert the given record in the table, returns error if no-new
+// record is given. The record id is set if it's empty.
+func (s *Store) Insert(record Record) error {
+	if !record.IsPersisted() {
```
Comment (dpordomingo) [inline store.go]: isNew != isPersisted
then... it should be 
`if record.IsPersisted() {
    return ErrNonNewDocument
}`

File: store.go [hunk 2]
```diff
@@ -1,199 +1,207 @@
...
-func (s *Store) Save(doc DocumentBase) (updated bool, err error) {
-	id := doc.GetId()
-	if len(id) == 0 {
-		return false, ErrEmptyID
+// Update updates the given fields of a record in the table. All fields are
+// updated if no fields are provided. For an update to take place, the record is
+// required to have a non-empty ID and not to be a new record.
+// Returns the number of updated rows and an error, if any.
+func (s *Store) Update(record Record, cols ...string) (int64, error) {
+	if record.IsPersisted() {
```
Comment (dpordomingo) [inline store.go]: isNew != isPersisted
then... it should be 
`if !record.IsPersisted() {
    return 0, ErrNewDocument
}`

File: store.go [hunk 3]
```diff
@@ -1,199 +1,207 @@
...
+		Exec()
+	if err != nil {
+		return err
 	}
 
-	sess, c := s.getSessionAndCollection()
-	defer sess.Close()
-
-	return c.UpdateId(doc.GetId(), doc)
+	record.setPersisted(false)
```
Comment (dpordomingo) [inline store.go]: isNew != isPersisted
then... it should be 
`record.setPersisted(true)`

File: store.go [hunk 4]
```diff
@@ -1,199 +1,207 @@
...
+// Save inserts or updates the given record in the table. It requires a record
+// with non empty ID.
+func (s *Store) Save(record Record) (updated bool, err error) {
+	if record.GetID().IsEmpty() {
+		return false, ErrEmptyID
 	}
 
-	if q.GetLimit() != 0 {
-		mq.Limit(q.GetLimit())
+	if record.IsPersisted() {
```
Comment (dpordomingo) [inline store.go]: isNew != isPersisted
then... it should be 
`if !record.IsPersisted() {return false, s.Insert(record)}`

**Comment by erizocosmico** [comment]
Fixed @dpordomingo

File: model.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
+package kallax
+
+import (
+	"database/sql"
+	"database/sql/driver"
+
+	uuid "github.com/satori/go.uuid"
+)
+
+type ID uuid.UUID
```
Comment (Serabe) [inline model.go]: Is this the _de facto_ package for UUID?

File: model.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+func (id ID) IsEmpty() bool {
+	return uuid.UUID(id) == uuid.Nil
+}
+
+type Model struct {
+	ID        ID
+	persisted bool
+}
+
+func (m *Model) GetID() ID {
```
Comment (Serabe) [inline model.go]: Isn't the idiom for getters to just be the name of the property?

File: model.go [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+func (m *Model) SetID(id ID) {
+	m.ID = id
+}
+
+func (m *Model) IsPersisted() bool {
+	return m.persisted
+}
+
+func (m *Model) setPersisted(isPersisted bool) {
+	m.persisted = isPersisted
```
Comment (Serabe) [inline model.go]: I would say that something that is persisted cannot be set as not persisted, so maybe just `setPersisted` without params.

File: operators.go [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+func And(conds ...Condition) Condition {
+	return and{squirrel.And(condsToSqlizers(conds))}
+}
+
+// Not returns the given condition negated.
+func Not(cond Condition) Condition {
+	return not{cond}
+}
+
+// Lt returns a condition that will be true when `col` is lower than `value`.
```
Comment (Serabe) [inline operators.go]: Wrong documentation.

File: query.go [hunk 9]
```diff
@@ -1,106 +1,203 @@
...
+			Select().
+			From(table),
+		batchSize: 50,
 	}
+}
 
-	return operators.And(q.clauses...)
+// Select adds the given columns to the list of selected columns in the query.
+func (q *BaseQuery) Select(columns ...string) {
+	q.columns.add(columns...)
```
Comment (Serabe) [inline query.go]: Shouldn't `Select` remove the columns from the excluded set as well?

Think of:

```go
query.SelectNot('name')
// In other place
query.Select('name')
```

User should be able to select columns even if those were excluded before.

File: query.go [hunk 10]
```diff
@@ -1,106 +1,203 @@
...
 }
 
-// Sort sets the sorting cristeria of the query.
-func (q *BaseQuery) Sort(s Sort) {
-	q.sort = s
+// SelectNot adds the given columns to the list of excluded columns in the query.
+func (q *BaseQuery) SelectNot(columns ...string) {
+	q.excludedColumns.add(columns...)
```
Comment (Serabe) [inline query.go]: See comment before. Though the usual processing steps would include a `select - excluded` algorithm, that would tie two methods, adding an implicit dependency between them.

File: query.go [hunk 11]
```diff
@@ -1,106 +1,203 @@
...
+	q.builder = q.builder.Where(squirrel.Sqlizer(cond))
+}
+
+// compile returns the selected column names and the select builder.
+func (q *BaseQuery) compile() ([]string, squirrel.SelectBuilder) {
+	columns := q.selectedColumns()
+	return columns, q.builder.Columns(columns...)
+}
+
+// String returns the SQL generated by the
```
Comment (Serabe) [inline query.go]: `by the`...

File: query_test.go [hunk 12]
```diff
@@ -1,27 +1,77 @@
...
+
+	s.Equal(s.q, copy)
+	s.NotEqual(unsafe.Pointer(s.q), unsafe.Pointer(copy))
+}
+
+func (s *QuerySuite) TestSelectedColumns() {
+	s.q.Select("a", "b", "c")
+	s.q.SelectNot("b")
+	s.Equal([]string{"a", "c"}, s.q.selectedColumns())
+}
```
Comment (Serabe) [inline query_test.go]: Add tests for `Select + SelectNot + Select` where at least one of the fields is passed to the three calls. Same for `SelectNot + Select + SelectNot`.

File: resultset.go [hunk 13]
```diff
@@ -1,87 +1,54 @@
...
-	//ErrNotFound document not found
-	ErrNotFound = errors.New("document not found")
-	// ErrStop if is used on a callback of a ResultSet.ForEach function the loop
-	// is stopped
-	ErrStop = errors.New("document not found")
-)
+// ErrRawScan is an error returned when a the `Scan` method of `ResultSet`
+// is called with a `ResultSet` created as a result of a `RawQuery`, which is
+// not allowed.
+var ErrRawScan = fmt.Errorf("result set comes from raw query, use RawScan instead")
```
Comment (Serabe) [inline resultset.go]: Shouldn't we use Máximo's package?

File: store.go [hunk 14]
```diff
@@ -1,199 +1,207 @@
...
+	result, err := s.updater.
+		SetMap(clauses).
+		Where(squirrel.Eq{s.schema.Identifier(): record.GetID()}).
+		Exec()
+	if err != nil {
+		return 0, err
+	}
 
-	return c.RemoveId(doc.GetId())
+	return result.RowsAffected()
```
Comment (Serabe) [inline store.go]: Can this be other than 0 or 1? If it is 0, should'n that be an error?

File: model.go (hunk 5 above)
Comment (dpordomingo) [inline model.go]: I found it as the best uuid package too

File: model.go (hunk 6 above)
Comment (dpordomingo) [inline model.go]: using ID() would conflict with the homonym ID property

File: model.go (hunk 7 above)
Comment (dpordomingo) [inline model.go]: I did it in my branch, that I'll pullrequest during the day :)

File: model.go (hunk 5 above)
Comment (erizocosmico) [inline model.go]: yep

File: model.go (hunk 7 above)
Comment (erizocosmico) [inline model.go]: yep, this is not a final implementation of model

File: query.go (hunk 9 above)
Comment (erizocosmico) [inline query.go]: totally right

File: query.go (hunk 11 above)
Comment (erizocosmico) [inline query.go]: lol

File: resultset.go (hunk 13 above)
Comment (erizocosmico) [inline resultset.go]: is it ready yet?

File: store.go (hunk 14 above)
Comment (erizocosmico) [inline store.go]: Yep, maybe it should

**Comment by erizocosmico** [comment]
@Serabe we'll add the testing thingy later to the README along with all documentation

**Comment by erizocosmico** [comment]
Fixed @Serabe

File: resultset.go (hunk 13 above)
Comment (Serabe) [inline resultset.go]: Not sure :$

**Pull request merged**